cli/src/commands/test/gas_report.rs

```diff
@@ -0,0 +1,158 @@
+    modifiers::UTF8_ROUND_CORNERS, presets::UTF8_FULL, Attribute, Cell, Color, Table,
+};
+use kythera_lib::{ExecutionEvent, Method, TestResult, TestResultType, WasmActor};
+
+/// Gas report for the tested contracts.
+#[derive(Default, Debug)]
+pub struct GasReport<'a> {
+    reports: BTreeMap<&'a WasmActor, ActorInfo<'a>>,
+}
+
+#[derive(Debug, Default)]
+/// Actor method calls information
+/// TODO: calculate actor deployment gas consumption.
+pub struct ActorInfo<'a> {
+    methods: BTreeMap<&'a Method, Vec<u64>>,
+}
+
+impl<'a> GasReport<'a> {
+    /// Analyze a set of [`TestResult`]s for a target Actor.
+    pub fn analyze(&mut self, actor: &'a WasmActor, test_results: &[&TestResult]) {
+        let mut info = match self.reports.remove(actor) {
+            Some(info) => info,
+            None => ActorInfo::default(),
+        };
+
+        for result in test_results {
+            let apply_ret = match result.ret() {
+                TestResultType::Passed(apply_ret) | TestResultType::Failed(apply_ret) => apply_ret,
+                TestResultType::Erred(_) => {
+                    continue;
+                }
+            };
+
+            // Get the Gas consumption by each Call of the Target actor.
+            let mut stack = vec![];
+            for trace in &apply_ret.exec_trace {
+                match trace {
+                    ExecutionEvent::GasCharge(gas_charge) => {
+                        // Add this gas charge to the total gas charge by the main method.
+                        // There is a `GasCharge` before any `Call` so we skip it.
+                        let (_method, total_gas_charge) = match stack.get_mut(0) {
+                            Some(e) => e,
+                            None => continue,
+                        };
+                        *total_gas_charge =
+                            *total_gas_charge + gas_charge.compute_gas.as_milligas();
+                    }
+
+                    ExecutionEvent::Call { method, .. } => {
+                        stack.push((method, 0));
+                    }
+                    ExecutionEvent::CallReturn(_, _) | ExecutionEvent::CallError(_) => {
+                        let (method_num, total_gas_charge) =
+                            stack.pop().expect("A call return should match a Call");
+                        // If stack is empty we reached the main Method call.
+                        if !stack.is_empty() {
+                            continue;
+                        }
+                        // If the method called was from the target actor,
+                        // we create a new call on `GasInfo` with the totals of gas charge.
+                        let Some(method) = actor
+                                .abi()
+                                .methods()
+                                .iter()
+                                .find(|a| a.number() == *method_num) else {
+                            continue;
+                        };
```

Comment:
Only checking method number is not enough. We should also check that the call has something to do with our deployed target actor.

Reply:
Method number is _in fine_ only a hash of a name. A name could be used across multiple actors, even one that is not our deployed target actor. Therefore the pair `(method_number, target_actor_id)` is unique to identify a method that is interesting to us but `method_number` is not.